From jfs-patches-admin@www-124.ibm.com Wed Apr 14 19:47:35 2004
Return-Path: <agruen@suse.de>
Received: from imap-dhs.suse.de ([unix socket]) by imap-dhs (Cyrus v2.1.16)
	with LMTP; Wed, 14 Apr 2004 19:47:35 +0200
X-Sieve: CMU Sieve 2.2
Received: from hermes.suse.de (hermes.suse.de [10.0.0.1]) (using TLSv1 with
	cipher EDH-RSA-DES-CBC3-SHA (168/168 bits)) (Client CN "hermes.suse.de",
	Issuer "SuSE Linux AG internal IMAP-Server CA" (verified OK)) by
	imap-dhs.suse.de (Postfix) with ESMTP id 13E364D8455 for
	<agruen@imap-dhs.suse.de>; Wed, 14 Apr 2004 19:47:35 +0200 (CEST)
Received: from wotan.suse.de (wotan.suse.de [10.10.0.1]) by hermes.suse.de
	(Postfix) with ESMTP id 1905A35FB6 for <agruen@imap-dhs.suse.de>; Wed, 14
	Apr 2004 19:47:36 +0200 (CEST)
Received: by wotan.suse.de (Postfix, from userid 10268) id 13FCF13601; Wed,
	14 Apr 2004 19:47:36 +0200 (CEST)
Received: from hermes.suse.de (hermes.suse.de [10.0.0.1]) by wotan.suse.de
	(Postfix) with ESMTP id EFF9F10C1A for <agruen@wotan.suse.de>; Wed, 14 Apr
	2004 19:47:35 +0200 (CEST)
Received: by hermes.suse.de (Postfix) id E8E3C35FD3; Wed, 14 Apr 2004
	19:47:35 +0200 (CEST)
Received: from scanhost.suse.de (scanhost.suse.de [10.0.0.5]) by
	hermes.suse.de (Postfix) with ESMTP id DBE0A35FB6; Wed, 14 Apr 2004
	19:47:35 +0200 (CEST)
Received: from hermes.suse.de ([10.0.0.1]) by scanhost.suse.de (scanhost
	[10.0.0.5]) (amavisd-new, port 10025) with ESMTP id 14906-06; Wed, 14 Apr
	2004 19:47:30 +0200 (CEST)
Received: from Cantor.suse.de (ns.suse.de [195.135.220.2]) (using TLSv1
	with cipher EDH-RSA-DES-CBC3-SHA (168/168 bits)) (No client certificate
	requested) by hermes.suse.de (Postfix) with ESMTP id C957836211; Wed, 14
	Apr 2004 19:46:59 +0200 (CEST)
Received: from oss.software.ibm.com (www-126.ibm.com [129.33.28.113]) by
	Cantor.suse.de (Postfix) with ESMTP id CC910459D00; Wed, 14 Apr 2004
	19:46:21 +0200 (CEST)
Received: from localhost ([127.0.0.1] helo=swg3ws026.southbury.usf.ibm.com)
	by oss.software.ibm.com with esmtp (Exim 3.34 #1) id 1BDoSl-0000KO-00; Wed,
	14 Apr 2004 13:46:15 -0400
Received: from kleikamp.dyn.webahead.ibm.com (pixpat.austin.ibm.com
	[192.35.232.241]) by www-126.ibm.com (8.11.6/8.11.0) with ESMTP id
	i3EHjoU01220 for <jfs-patches@www-124.southbury.usf.ibm.com>; Wed, 14 Apr
	2004 13:45:50 -0400
Received: from kleikamp.dyn.webahead.ibm.com (localhost [127.0.0.1]) by
	kleikamp.dyn.webahead.ibm.com (Postfix) with ESMTP id A879121C085 for
	<jfs-patches@www-124.southbury.usf.ibm.com>; Wed, 14 Apr 2004 12:45:44
	-0500 (CDT)
Received: (from shaggy@localhost) by kleikamp.dyn.webahead.ibm.com
	(8.12.8/8.12.8/Submit) id i3EHjiSk005689 for
	jfs-patches@www-124.southbury.usf.ibm.com; Wed, 14 Apr 2004 12:45:44 -0500
From: shaggy@austin.ibm.com
Message-Id: <200404141745.i3EHjiSk005689@kleikamp.dyn.webahead.ibm.com>
To: jfs-patches@oss.software.ibm.com
Subject: [Jfs-patches] [PATCH] Fix race in jfs_sync (1 of 1)
Sender: jfs-patches-admin@www-124.ibm.com
Errors-To: jfs-patches-admin@www-124.ibm.com
X-BeenThere: jfs-patches@www-124.ibm.com
X-Mailman-Version: 2.0.4
Precedence: bulk
List-Unsubscribe:
	<http://www-124.ibm.com/developerworks/oss/mailman/listinfo/jfs-patches>,
	<mailto:jfs-patches-request@www-124.ibm.com?subject=unsubscribe>
List-Id: <jfs-patches.www-124.ibm.com>
List-Post: <mailto:jfs-patches@www-124.ibm.com>
List-Help: <mailto:jfs-patches-request@www-124.ibm.com?subject=help>
List-Subscribe:
	<http://www-124.ibm.com/developerworks/oss/mailman/listinfo/jfs-patches>,
	<mailto:jfs-patches-request@www-124.ibm.com?subject=subscribe>
List-Archive: <http://www-124.ibm.com/pipermail/jfs-patches/>
Date: Wed, 14 Apr 2004 12:45:44 -0500
X-Virus-Scanned: by amavisd-new at scanhost.suse.de
X-Spam-Status: No, hits=-4.7 tagged_above=-20.0 required=5.0
	tests=BAYES_00, NO_REAL_NAME
X-Spam-Level: 
Content-Type: text/plain; CHARSET=us-ascii
X-Evolution-Source: imap://agruen@imap-dhs.suse.de/
Content-Transfer-Encoding: 8bit
Mime-Version: 1.0

# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2004/04/12 13:17:58-05:00 shaggy@austin.ibm.com 
#   JFS: Fix race in jfs_sync
#   
#   Don't let final iput happen while jfs_sync is processing inode.
# 
diff -Nru a/fs/jfs/jfs_txnmgr.c b/fs/jfs/jfs_txnmgr.c
--- a/fs/jfs/jfs_txnmgr.c	Wed Apr 14 12:44:33 2004
+++ b/fs/jfs/jfs_txnmgr.c	Wed Apr 14 12:44:33 2004
@@ -2977,11 +2977,12 @@
 					    anon_inode_list);
 			ip = &jfs_ip->vfs_inode;
 
-			/*
-			 * down_trylock returns 0 on success.  This is
-			 * inconsistent with spin_trylock.
-			 */
-			if (! down_trylock(&jfs_ip->commit_sem)) {
+			if (! igrab(ip)) {
+				/*
+				 * Inode is being freed
+				 */
+				list_del_init(&jfs_ip->anon_inode_list);
+			} else if (! down_trylock(&jfs_ip->commit_sem)) {
 				/*
 				 * inode will be removed from anonymous list
 				 * when it is committed
@@ -2991,6 +2992,8 @@
 				rc = txCommit(tid, 1, &ip, 0);
 				txEnd(tid);
 				up(&jfs_ip->commit_sem);
+
+				iput(ip);
 				/*
 				 * Just to be safe.  I don't know how
 				 * long we can run without blocking
@@ -3010,6 +3013,10 @@
 				/* Put on anon_list2 */
 				list_add(&jfs_ip->anon_inode_list,
 					 &TxAnchor.anon_list2);
+
+				TXN_UNLOCK();
+				iput(ip);
+				TXN_LOCK();
 			}
 		}
 		/* Add anon_list2 back to anon_list */
_______________________________________________
Jfs-patches mailing list
Jfs-patches@www-124.ibm.com
http://www-124.ibm.com/developerworks/oss/mailman/listinfo/jfs-patches

